docs: add interactive environment docs (#980)
SwissDataScienceCenter/renku: docs/user/interactive_environments.rst

.. _interactive_environments:

Interactive Environments on RenkuLab
====================================

RenkuLab provides convenient interactive computing environments based on
`JupyterLab <https://jupyterlab.readthedocs.io/en/stable/>`_. You can use them
to quickly access your project in fully-configured Python, RStudio or Julia
sessions. They make it easy to collaborate by providing reproducible
computational sessions for everyone working together on a project. They can
even be used to host interactive dashboards running on live data for use

> in demos or presentations.

Reviewer: I applied this change but I think I will remove it - this is supposed to be very high-level text and talking about pushing to GitLab seems a bit out of scope. We make that point more explicit in other parts of the docs.

* ``Basic Python Project``: JupyterLab
* ``Basic R Project``: RStudio AND JupyterLab with R kernel
* ``Minimal Renku`` (language-agnostic): JupyterLab

This can be customized -- see :ref:`customizable`.

Starting a new Interactive Environment
--------------------------------------

When you choose to start a new interactive environment, you will need to select
some configurations. If you're not sure what options to select, it's best to keep
the defaults. If, however, you've tried launching a project with the defaults and
it crashed on you, you might want to increase some processing power or memory.

Here's the run down for the configuration options.

Option        Description
============= =======================================================================
branch        default master, but if you're doing work on another branch, switch!
commit        default latest, but you can launch the environment from an earlier
              commit; also useful if your latest commit's build failed (see below).
environment   ``lab``: JupyterLab; ``rstudio``: RStudio; if you're using a python
              template, the ``rstudio`` endpoint will not work.

> # CPUs        the number of CPUs available; select the lowest # available unless

Reviewer: should it be 

> CPUs       the amount of CPU requested for your session

              strings that are not your data, run ``renku storage pull <filepath>``
              to get the relevant files, or ``git lfs pull`` to get all of the
              files at once.

What if the Dockerfile isn't available?
---------------------------------------

When you select the configurations for ``branch`` and ``commit``, a check is run
to see if there's an image for this state available in the project's image registry,
which is stored in Renkulab's instance of GitLab. The container is then launched
from that built image.

An image build from the ``Dockerfile`` in the project is kicked off automatically
using GitLab's CI/CD configured by the project's ``.gitlab-ci.yml`` when you:

 * create the project
 * fork a project (in which the new build happens for the fork)
 * push changes to the project

It can sometimes take a long time to build for various reasons, but if you've just
created the project on Renkulab from one of the templates it should take less than
a minute.

The Dockerfile is still building
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If the ``Dockerfile`` has a "still building" message, you can either wait patiently,

> or watch it build by clicking associated link to see the streaming log messages

Reviewer: nice work article police, that's usually my job :)

check out :ref:`Customizing`

Saving your work
----------------

Interactive environments persist for 24 hours. If left unattended, they might shut
down without warning. They also might crash if you run a process that eats more
memory than you've allocated. Thus, it's best to save often.

There are two ways to save your work back to Renkulab from an Interactive Environment
(both available in JupyterLab and RStudio), and behind the scenes both are using ``git``
staging (``add``), ``commit``, and ``push``. You can type these commands directly
into the available terminal interface of your Interactive Environment, or click
some buttons via the git plugins.

When you push your changes back to Renkulab, the GitLab CI/CD is triggered to build
a new image out of the ``Dockerfile``, which will be available the next time you
start a new environment.

Saving via Terminal
~~~~~~~~~~~~~~~~~~~

In the Terminal interface inside the Interactive environment, call the following
three ``git`` commands to stage, commit, and push your changes:

1. ``git add *``
2. ``git commit -m "my short but descriptive message of the changes I made"``

> 3. ``git push``
> 
> Saving via Git Plugin

Reviewer: ```suggestion
3. ``git push``

If you are new to git, these resources might be useful:

* [git documentation](https://git-scm.com/doc)
* [A great interactive cheatsheet](http://ndpsoftware.com/git-cheatsheet.html)
* [An intro to git](https://rogerdudler.github.io/git-guide/)

Saving via Git Plugin
```

tab on the Renkulab UI.

Sometimes an environment will get stuck pending launch (e.g. because you requested
GPUs and they are not yet available). In this case you can view the status and
logs to see if there's a useful message. In the case of pending GPUs, you can
safely wait for the resources to become available. However in other cases, more
commonly when you are trying to customize your environment, you might notice an
issue via the logs, and want to stop the launch so that you can fix the problem
and try again. However, for these stuck notebooks it is not yet possible.
You can reach out to us on `discourse <https://renku.discourse.group>`_ in this
case.

Reproducibility
---------------

What are the limitations of the Interactive Environments?
---------------------------------------------------------

How can I customize my Interactive Environment?
-----------------------------------------------

The launch is enabled by the content in the following files in your project:

* ``Dockerfile``: defines the type of interactive environment and other software
  installed in the environment, including the ``renku`` command-line installation.

> * ``.gitlab-ci.yml``: defines the docker build of the image based on the project's

Reviewer 2: ```suggestion
* ``.gitlab-ci.yml``: controls the docker build of the image based on the project's
```

  ``Dockerfile``.

Modifying the template's Dockerfile
-----------------------------------

The `Dockerfile` in the project is what defines the environment. In the template
provided, the `Dockerfile`

> Renku Project uses Docker for containerization, which While we

Reviewer 2: ```suggestion
Renku Projects use Docker for containerization. While we
```

> have a set of defaults that we build into a minimal Python and R images

Reviewer 2: ```suggestion
have a set of defaults that we build into a minimal Python and R image
```

image, there are several reasons why you might want to build on top of these or
write your own entirely.

Dockerfile structure
~~~~~~~~~~~~~~~~~~~~

The project's ``Dockerfile`` lives in the top level of the project directory. In
the default ``Dockerfile`` provided in the template, the first line is a ``FROM``
statement that specifies a `versioned base docker image <https://github.com/SwissDataScienceCenter/renku-jupyter>`_.
We add new versions periodically, but the heart of it is the set of installations
of jupyterlab/rstudio, git, and renku::

  FROM renku/singleuser:0.3.5-renku0.5.2

  # or, for RStudio in the build

  FROM renku/singleuser-r:0.3.5-renku0.5.2

The next two statements install user-specified libraries from ``environment.yml``
and ``requirements.txt``::

  # install the python dependencies
  COPY requirements.txt environment.yml /tmp/
  RUN conda env update -q -f /tmp/environment.yml && \
  /opt/conda/bin/pip install -r /tmp/requirements.txt && \
  conda clean -y --all && \
  conda env export -n "root"

You can add to this ``Dockerfile`` in any way you'd like.

.. _docker_dev:

Dockerfile development
~~~~~~~~~~~~~~~~~~~~~~

> If you're going to be making simple modifications to the ``Dockerfile`` (i.e. changing

Reviewer 2: Maybe some Docker references? I think the official [Dockerfile reference](https://docs.docker.com/engine/reference/builder/) is quite ok.

the base Docker image version number), you can use the following steps to update
and re-build the image:

#. On the project's landing page, click the **View in GitLab** button in the upper righthand corner (opens a new tab by default).
#. In GitLab, click the **Repository** tab in the lefthand column, which drops you into the **Files** tab.
#. Click the **Dockerfile** out of the list of files that appears, and click **Edit** (top right, near the red Delete button. Don't click the red Delete button.)
#. Make your edits in this window.
#. When you're satisfied with the edits, scroll down and write a meaningful **commit message** (you'll thank yourself later).

> #. Click the green **Commit changes** button.
> 
> Now you have committed the changes to your ``Dockerfile``. Since you have made a commit,

Reviewer: ```suggestion
#. Click the green **Commit changes** button.

You may find the [official docker documentation](https://docs.docker.com/engine/reference/builder/) useful 
during this process.

Now you have committed the changes to your ``Dockerfile``. Since you have made a commit,
```

the CI/CD pipeline will kick off (pre-configured for you as a ``renkulab-runner``

> inside the GitLab CI/CD settings). It will attempt to rebuild your project with

Reviewer 2: The runner is the virtual machine executing the CI jobs, but I think that's more of a concern for admins than for users, so I'd just leave that away.

the new contents of your ``Dockerfile`` based on the configuration in ``.gitlab-ci.yml``,
a file at the top level of your project directory.

> The contents of ``.gitlab-ci.yml`` show you that in the build stage, we pull

Reviewer 2: I think we should add here that normally people will not have to edit their `.gitlab-ci.yaml`.

the docker image for Renku, build our new image out of our ``Dockerfile``
with a tag relating to the commit, and push it.

Let's monitor this process:

#. Click the **CI/CD** > **Jobs** tab.
#. Click the latest status that corresponds to the changes to the ``Dockerfile`` you just made (probably "running", unless it's already "completed" or "failed").

This is the log file from the build process specified in the ``.gitlab-ci.yml``
file. If it succeeds, there will be a green **passed** status, and the end of
log will be a green **Job succeeded**. If the build instead failed, you can use
the messages in the log to determine why and hint at what you can do to fix it.

.. warning::

  Note that the settings have been configured for this build to time out and fail
  after one hour. While a long running build might be indicative of a bug in your
  ``Dockerfile``, it's possible that your build might take a long time. If this is the

>   case, you can change these settings via the lefthand column of the GitLab

Reviewer: ```suggestion
  case, you can change the limits in the project settings via the lefthand column of the GitLab
```

>   interface at **Settings** > **CI/CD** > **General pipelines** > **Timeout**.

Reviewer 2: Maybe distinguish project settings from user settings here.

> Using your new Dockerfile

Reviewer 2: ```suggestion
Using your new Docker image
```

~~~~~~~~~~~~~~~~~~~~~~~~~

Passing CI/CD is great, but in order to use the new image you need to
(re)start your interactive environment.

To do this, go back to the Renku platform, and from the project's landing page,
first check in the **Files** tab that your changes to the ``Dockerfile`` are
present. If not, you can force-refresh the page. Then, go to the **Notebook

> servers** tab. If you have any running notebooks, those will remain built with

Reviewer 2: ```suggestion
servers** tab. If you have any running notebooks, those will keep running the image which was built with
```

the older version(s) of the ``Dockerfile``. You can **Start new server** and
**Launch server** to start a notebook with the latest image.

If the server launches, test it to make sure that the extra functionality you
added in the ``Dockerfile`` is present in the container. If it is not, you can
go back to the GitLab interface and continue to make changes until you are
satisfied.

Looking to make more extensive modifications? Build running too long? Check out
the `next section <_more_extensive_docker>`_.

.. _more_extensive_docker:

> For more extensive modifications

Reviewer 2: ```suggestion
for more extensive modifications.

More extensive modifications
```

get set up and learn how to build and test local images.

Once you have a local docker setup, you can clone your project locally (if you
haven't set up an SSH key from GitLab you'll need to do this), make
modifications to the ``Dockerfile``, and ``docker build`` and ``docker run`` to
test your changes. To test whether your docker image will work, try running it
with::

  docker run --rm -ti -p 8888:8888 <image> jupyter lab --ip=0.0.0.0

.. warning::

  You need to install ``jupyter`` and ``jupyterhub`` into the image to be able to
  start notebook servers on renkulab.io.

You can commit these changes and push to the repo. Then, follow the rest of the
steps in :ref:`docker_dev`.

Note that by default there are two choices for the ``Dockerfile`` (chosen at
project creation time via "python base" or "R base") for the base image, located
here:

* a `JupyterLab base <https://github.com/SwissDataScienceCenter/renku-jupyter/tree/master/docker/base>`_ (with renku installed on top)
* a `rocker (R + RStudio) base <https://github.com/SwissDataScienceCenter/renku-jupyter/tree/master/docker/r>`_ (with conda and renku installed on top)

> These two images are pushed into `dockerhub <https://hub.docker.com/r/renku/>`_.

Reviewer 2: ```suggestion
These two images are available on `dockerhub <https://hub.docker.com/r/renku/>`_.
```
